virtualizarr/readers/safetensors.py

```diff
@@ -0,0 +1,641 @@
+import json
+import struct
+from collections.abc import Iterable, Mapping
+from typing import Any, Dict, Optional
+from urllib.parse import urlparse
+
+import numpy as np
+from obstore.store import (
+    HTTPStore,
+    ObjectStore,  # type: ignore[import-not-found]
+)
+from xarray import Dataset, Index
+
+from virtualizarr.manifests import (
+    ChunkEntry,
+    ChunkManifest,
+    ManifestArray,
+    ManifestGroup,
+    ManifestStore,
+    ObjectStoreRegistry,
+)
+from virtualizarr.manifests.store import default_object_store
+from virtualizarr.manifests.utils import create_v3_array_metadata
+from virtualizarr.readers.api import VirtualBackend
+from virtualizarr.types import ChunkKey
+
+
+class SafeTensorsVirtualBackend(VirtualBackend):
+    """
+    Backend for reading SafeTensors files as virtual datasets.
+
+    SafeTensors is a format for safely storing tensors (multidimensional arrays),
+    without using pickle, and with zero-copy access. It is commonly used for storing
+    model weights in the fields of ML and AI.
+
+    The format consists of:
+    - 8 bytes (header size): unsigned little-endian 64-bit integer containing the size of the header
+    - N bytes (header): a JSON UTF-8 string containing tensor metadata
+    - Rest of the file: byte-buffer containing tensor data
+
+    Examples
+    --------
+    Open a local SafeTensors file with default settings:
+
+    >>> vds = open_virtual_dataset("model_weights.safetensors")
+
+    Open a SafeTensors file with custom dimension names:
+
+    >>> custom_dims = {"weight": ["input_dims", "output_dims"], "bias": ["output_dims"]}
+    >>> vds = open_virtual_dataset(
+    ...     "model_weights.safetensors",
+    ...     virtual_backend_kwargs={"dimension_names": custom_dims}
+    ... )
+
+    Open a GPT-2 model from Hugging Face Hub:
+
+    >>> vds = open_virtual_dataset(
+    ...     "https://huggingface.co/openai-community/gpt2/resolve/main/model.safetensors"
+    ... )
+    >>> # Access various GPT-2 tensors
+    >>> print(vds["wte.weight"].shape)  # Word Token Embeddings: (50257, 768)
```

Comment:
What's the structure of these variable names? If they are nested behind a dot separator syntax I wonder if it would be helpful to return an xarray.DataTree instead...

Reply:
Yes, this is an excellent question.

Technically, these are hierarchical structures, and so would seem perfectly suited to Zarr groups. So, for example `a.b.c` represents 
```
└── a/
    └── b/
        └── c/
```

However, the reality is that `pytorch` is a dominant framework for training/inference of these models, and it has the concept of a [state dict](https://pytorch.org/tutorials/recipes/recipes/what_is_state_dict.html) for loading model weights, and it expects a flattened structure i.e. the current `a.b.c` dictionary key layout.

I'm not enough of an expert to know what the best option is here, and so I have gone with the flattened structure, to ease integration with `pytorch`.

I'll leave a comment about this.